### PR TITLE
refactor particle manipulators/filter/startPosition

### PR DESCRIPTION
**Pull request by psychocoderHPC**
# PIConGPU
- create functor folder
  - move `manipulators/generic/detail/Rng.*` to `functor/misc`
  - add functor `User` to reduce code duplication for the optional current step within the constructor
- filter: add `IUnary` interface for filter to secure the interfaces
- plugin: enforce in all plugins which uses filter that the interface is fulfilled
- manipulator:
  - update used dependency `Rng`
  - add documentation
  - set `getName()` static
  - use `misc::User` to initialize the user functor

# PMacc
- make `getName()` static to allow working with the name withou creating a instance
- remove return type detection (simplefy the code)
**Note:** There is no change in the algorithmic part of the methods.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
please wait with merging, I am working on the next refactoring. Maybe we should join both to minimize the diff.

**Comment by ax3l** [comment]
Sure, no problem. Can you please add a [WIP] to the title?

**Comment by psychocoderHPC** [comment]
ok this PR is now ready: please review the PR commit by commit

File: include/picongpu/particles/manipulators/generic/FreeRng.def [hunk 1]
```diff
@@ -74,7 +76,7 @@ namespace generic
     >
     struct FreeRng;
 
-} // namespace generic
+} // namepsace generic
```
Comment (ax3l) [inline include/picongpu/particles/manipulators/generic/FreeRng.def]: typo introduced :)

File: include/picongpu/particles/filter/filter.hpp [hunk 2]
```diff
@@ -19,5 +19,6 @@
 
 #pragma once
 
+
```
Comment (ax3l) [inline include/picongpu/particles/filter/filter.hpp]: pls do not add a double newline before & after `#pragma once` but a double newline after the last include

File: include/picongpu/particles/functor/User.def [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
```
Comment (ax3l) [inline include/picongpu/particles/functor/User.def]: here same

File: include/picongpu/particles/functor/User.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
```
Comment (ax3l) [inline include/picongpu/particles/functor/User.hpp]: here same

File: include/picongpu/particles/functor/functor.def [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
```
Comment (ax3l) [inline include/picongpu/particles/functor/functor.def]: here same

File: include/picongpu/particles/functor/functor.hpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
```
Comment (ax3l) [inline include/picongpu/particles/functor/functor.hpp]: here same

File: include/picongpu/particles/manipulators/generic/FreeRng.def [hunk 7]
```diff
@@ -19,6 +19,7 @@
 
 #pragma once
 
+
```
Comment (ax3l) [inline include/picongpu/particles/manipulators/generic/FreeRng.def]: here same

File: include/picongpu/particles/functor/User.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/picongpu/particles/functor/User.hpp]: fixed you commented the wrong line

**Comment by psychocoderHPC** [comment]
@ax3l I kill the cudafe++ for this PR the CI was freezing in the foil example. I pushed this PR again to trigger a new CI test.

File: include/picongpu/particles/functor/functor.hpp (hunk 6 above)
Comment (psychocoderHPC) [reply include/picongpu/particles/functor/functor.hpp]: fixed

**Comment by psychocoderHPC** [comment]
ready to merge

**Comment by ax3l** [comment]
I think we should not have merged this since CUDA 7.5 is running forever again 😢

**Comment by psychocoderHPC** [comment]
offline discussed with @ax3l : it could be that we break a magic number for the length of type names in CUDA 7.5.

**Pull request merged**